### ISSUE TITLE
[BUG] Extremely buggy. Keeps opening up Search box, pastes other person's text, and then stops because Facebook Search is broken.

### ISSUE THREAD

**Issue by gaveitatry2** [theahura/shoot-the-messenger#100]
So I click 'Start' and it will scroll up looking for messages to delete. But after a certain point, it will stop. For some reason it will open the "Search" box and paste a message that the other person sent (not me) in the Search box. It will show "No results" in the Search box, and if I click on the webpage, a popup appears that says "www.messenger.com says Unable to find message: (the message)". I've tried locating the message, which takes a lot of work scrolling up, and then deleting it. This worked the first time. But after that I keep running into new messages that it is trying to search for but can't find. And I feel like I've run into a wall and I can't use this extension. I think that the problem is that Facebook hasn't let me use the Search feature for a while now. Even without this extension, if I search for something, it can't find anything. Even if it is a recently sent message and I am looking at it. I don't know why the Facebook search (search inside the conversation) is not working for me. And I don't know if it is working for other people. I don't know how this extension works, but my guess is that it scrolls up and keeps deleting. But after a while, it wants to refresh, so it copies a person's message, refreshes the page, and then searches for the message. But since Facebook Search is broken (for me), it breaks the extension so that it won't be able to keep scrolling and deleting. The only way to fix this is figure out why Facebook Search isn't working (for me) or quit making the extension refresh, copy, search for, and paste. No more refreshes would probably solve my problem.

**Comment by gaveitatry2** [comment]
TL;DR

The ability to search inside a Facebook conversation for a message no longer works. Not in Mozilla Firefox, Google Chrome (facebook.com and messenger.com). Not even the Android Messenger App.

Because of this, the extension will quit working because it copies the other sender's text and then refreshes the page and then searches for the text that was copied.

So to fix this extension, one of two things needs to happen.

1.) Don't start deletes from the very bottom or end of the conversation. Start deletes from current position in the conversation. So if I go back to May 2023 and click "Start" start deleting from May 2023 and not from September 2023.

2.) Or just make the extension stop refreshing the page and stop searching for text in the conversation. Search no longer works and it breaks the extension.

**Comment by gaveitatry2** [comment]
Update: Facebook search is working for SOME conversations, but not this one. Maybe Facebook search does not work on very long conversations that were recently un-archived. I don't know. I search for stuff in some conversations and it works, and I search in other conversations and it won't work.

**Comment by 8SecSleeper** [comment]
You can disable the reloading functionality if it's a issue like in your
case.

I feel like there's a better way to handle the whole thing. Like just
removing the elements from memory as it progresses so there's no way to
reload it. However it's not my code base. I wrote patches to keep it
working previously. And I made the search use messages that the partner
sent so it wouldn't be searching for a message that doesn't exist anymore.

But I think that whole method could be scrapped as mentioned above.

**Comment by theahura** [comment]
@gaveitatry2 you basically nailed what's happening and why it's happening. Under the hood, the extension is only capable of doing what you as a logged in human can do -- if you can't search at all, it can't search at all. And you also nailed why we have that feature -- to clear out memory

Here's the issue (@8SecSleeper) there is no way to clear memory without refreshing the page. I tried a ton of different approaches here, but at the end of the day the extension is not fully privileged. It has access to the dom, but it does not have access to the underlying react code that is running the dom; there's no way to actually remove every reference to the deleted data in react, so the js garbage collector won't clear up deleted data, and eventually the system crashes.

Still, I agree this is the buggiest and hardest to maintain feature. Maybe we could improve this by having a toggle in the extension that disables the entire refresh loop 🤔

**Comment by 8SecSleeper** [comment]
That's unfortunate

**Comment by gaveitatry2** [comment]
@theahura 

**SUGGESTION**

What if you give us the option to also delete the other sender's messages? And if necessary also delete the "You have unsent a message." As an OPTION. That way, memory won't be an issue because the extension won't have to scroll back far to find the messages that have not been unsent yet.

**Comment by 8SecSleeper** [comment]
That's an option but will take quite awhile longer doing it that way do to
the rate limit we have to deal with.

**Comment by gaveitatry2** [comment]
I'd rather have an option that takes longer than no option at all (because that means I have way too many messages that I will never be able to unsend). At least with this option, I can unsent all my messages eventually without having to do it one by one, manually.

**Comment by gaveitatry2** [comment]
But my idea is to have TWO buttons:

1.) Unsend Your Messages
2.) Unsend Their Messages

This way, for those conversations that aren't too long, you can just use 1.) and it won't be as slow.
Also some people don't want to delete the other person's messages, so they will use 1.)
But for me, on my long conversation, I would start with 1.) Then when I encounter the "can't find message in search box", then I use 2.) and then when it catches up I switch back to 1.) and just keep taking turns between 1.) and 2.) until all messages are deleted.

**Comment by gaveitatry2** [comment]
> Maybe we could improve this by having a toggle in the extension that disables the entire refresh loop 🤔

That would be awesome too. I didn't see you post that.

**Comment by gaveitatry2** [comment]
This extension sucks. Sorry, but it is the truth. That conversation that I wanted to delete my messages on... it finally let me use the Search feature. So I tried using your extension to finish deleting my messages in that conversation. But it will scroll up a bit, refresh the page, and then do absolutely nothing. Or it will scroll up a bit, refresh the page, type something in the search box, and then do nothing. It doesn't even look like it presses enter after it types the message to search in the search box. So I decided to try manually pressing enter for the search, and it searched for one word "he" when the text in the search box was a few sentences long. I deleted some of the text in the search box so that I only kept the first eight words, and I pressed enter, but nothing came up.

I feel like this extension is just plain buggy. Even if Facebook Search worked 100% of the time, the extension still would only delete some messages and then after that it will fail to go past that and delete some more.

I think at least two things needs to happen before this extension can work. 1.) Stop making it refresh the page and 2.) Stop making it search for things. In addition to that, 3.) You could add a separate "delete their messages" option so that way there would be less messages to scroll back on. But I also think that 4.) There should be an option to delete the "You unsent a message" messages because those probably cause the extension to be glitchy too (too much scrolling).

**Comment by gaveitatry2** [comment]
![chrome_bFhYPTXjBB](https://github.com/theahura/shoot-the-messenger/assets/75463665/f572844a-fc76-48d4-b49f-7724b4e6b9a7)
This shows Search finally worked (but your extension still won't delete my messages).

![chrome_dPJiHL2Moo](https://github.com/theahura/shoot-the-messenger/assets/75463665/42695092-9772-4102-a73f-6c5c0b3d647a)
Other times, it scrolls up, then refreshes, then does nothing.
Or it scrolls up, refreshes, types something in search, but then doesn't even press enter and does nothing.
Or it scrolls up, and I get this error message.

**Comment by 8SecSleeper** [comment]
I'm curious why you didn't just do as instructed and disable the refresh
for the conversation you had trouble with.

As the owner of this extension stated. It's basically a no winner
situation. You can only do so much in a chrome extension. Deleting the
unsent part will simply make it take forever. You can code that part quite
easily if so desired.

Furthermore, You can use the open source python library to unsend messages
without any of the issues or limitations a chrome extension has.

**Comment by theahura** [comment]
> I feel like this extension is just plain buggy. Even if Facebook Search worked 100% of the time, the extension still would only delete some messages and then after that it will fail to go past that and delete some more.

I was able to replicate this, actually. I think fb changed some stuff on the messenger internals, so things are breaking more than usual. 

A few misc things on my todo list
- figure out why the extension is missing some messages while scrolling
- try and improve the search feature again
- add a 'delete all' mode that scrubs the user's messages, the 'unsent' messages, and everyone else's messages as a last resort

Not sure when I'll get to this though -- probably not this weekend since I have something that feels like the flu, but maybe next weekend

> This extension sucks. Sorry, but it is the truth.

@gaveitatry2 I understand that you're frustrated, and perhaps did not intend this, but this comes across as entitled and rude. Please do less of this

The people who contribute to this extension, myself included, are doing so because we enjoy it. We all have lives outside of this, we have no incentive to fix these problems. If you really want this fixed on _your_ timeline, you are welcome to either pay us for the time (there's a button in the extension and the readme!) or fix it yourself (it's open source! the entire extension is maybe 200 lines of code!). Speaking truthfully, I'm less inclined to fix this for you than I am for folks who emailed me with the same problem, and are much more respectful of my time. 

Again, totally ok if that wasn't the intended tone, but I'm just letting you know how it comes across

**Comment by gaveitatry2** [comment]
Hey, I liked everything you had to say until you started lecturing me and giving me that social justice warrior nonsense. I was originally disappointed because the extension didn't work due to Facebook searches not working. But when I found out that the extension STILL won't work even when Facebook searches are working, then I said something that was only completely true. I'm sorry I didn't water it down enough for you. I don't think saying that the program "sucks" is something you need to take so damn personally. It's just an honest review of the program's current state (and possibly future state). You don't need to call me entitled. I don't EXPECT anything. You don't need to tell me about your TIMELINE. I never said "you need to fix it right now". Sheesh. I was just saying the program DOES NOT WORK. I'm sorry I didn't use better words for snowflakes. Anyway, thank you for addressing the issues and for stating what you are thinking about doing about the issues. I look forward to you working on them and I hope you do get them worked out. And I'm not going to rush you. In fact, I won't post anymore. Since my wording seems to OFFEND and TRIGGER you. Have a good day.
